src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1031,6 +1031,15 @@ public static bool ContainsTypeParameters(this TypeSymbol type, HashSet<TypePara
         private static readonly Func<TypeSymbol, HashSet<TypeParameterSymbol>, bool, bool> s_containsTypeParametersPredicate =
             (type, parameters, unused) => type.TypeKind == TypeKind.TypeParameter && parameters.Contains((TypeParameterSymbol)type);
 
+        public static bool ContainsTypeParameterFromSymbol(this TypeSymbol type, Func<Symbol, bool> containingSymbolFilter)
```

Comment:
>ContainsTypeParameterFromSymbol [](http://example.com/codeflow?start=27&length=31)

It looks like there is only one usage of this helper. Generalization of this helper with a filter delegate doesn't feel necessary. I think a specialized helper targeting specific check we need will be prefered, something called "ContainsMethodTypeParameter" and the condition could be as simple as ```(type as TypeParameterSymbol)?.TypeParameterKind == TypeParameterKind.Method```

Reply:
I've made the method more specialised but I've left my implementation of the check (`ContainingSymbol is MethodSymbol` instead of `TypeParameterKind == Method`)

Not sure about correctness of my check if my check in terms of correct usage of code model, but the current implementation handles the case we discussed earlier. Consider the following:

Original code:
```
void GenericMethod<T>(T value)
{
    for (int i = 0; i < 1; i++)
    {
        DoStuff<T>(() => value);
    }
}

void DoStuff<T>(Func<T> lambda)
{
}
```
Generated code:
```
private void GenericMethod<T>(T value)
{
    DisplayClass<T> displayClass = new DisplayClass<T>();
    displayClass.value = value;
    for (int i = 0; i < 1; i++)
    {
        DoStuff(new Func(displayClass.Lambda));
    }
}
```
When the compiler chooses if to cache the delegate, the type parameter symbol has `TypeParameterKind == Method` (as it is in the original parameter symbol) but the container is an instance of `SynthesizedClosureEnvironment` as the parameter is moved to the synthesised display class.

Reply:
BTW, should I re-request a review via GitHub button when I think I fixed all the remarks? Or do you monitor all the updates made somehow else?

Reply:
>  should I re-request a review via GitHub button when I think I fixed all the remarks? Or do you monitor all the updates made somehow else?

There is no requirement to do this, however you may re-request a review when you feel appropriate. We don't use any special system to track active PRs other than what GitHub provides. So, if there are many active PRs, a PR might fall of the radar. When this happens, it is recommended to bring our attention to the PR.

Reply:
> I've left my implementation of the check (ContainingSymbol is MethodSymbol instead of TypeParameterKind == Method)

That is fine, both forms should be equivalent.

> When the compiler chooses if to cache the delegate, the type parameter symbol has TypeParameterKind == Method (as it is in the original parameter symbol) but the container is an instance of SynthesizedClosureEnvironment as the parameter is moved to the synthesised display class.

It sounds like you are saying that you have run into a case when a type type parameter has ```TypeParameterKind == TypeParameterKind.Method```. If that is the case, this is a bug.

Reply:
> It sounds like you are saying that you have run into a case when a type type parameter has TypeParameterKind == TypeParameterKind.Method. If that is the case, this is a bug.

Yep, that is the case I have run into.
It happens because the original type parameter is wrapped by `SynthesizedSubstitutedTypeParameterSymbol` ([code](https://github.com/dotnet/roslyn/blob/1cca63b/src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSubstitutedTypeParameterSymbol.cs#L18)) that passes some properties through the original type parameter symbol (e.g. `TypeParameterKind`), but some are replaced with a new implementation (e.g. `ContainingSymbol`).

I cannot judge if it is the correct behaviour, but if it is not the correct one we can create and issue and I can fix it. And to fix it I will need some help to decide what is the expected behaviour in such cases.

Reply:
> I cannot judge if it is the correct behaviour, but if it is not the correct one we can create an issue and I can fix it. But to fix it I will need some help to decide what is the expected behaviour in such cases.

I created a PR [https://github.com/dotnet/roslyn/pull/53968](https://github.com/dotnet/roslyn/pull/53968) with asserts that represent expected behavior. The asserts are failing for some scenarios. I think the fix for SynthesizedSubstitutedTypeParameterSymbol is to override TypeParameterKind property as it is done in SynthesizedClonedTypeParameterSymbol for VB.

Reply:
I'll use your branch and implement the fix.